File: sdk/ts/src/executor.ts
/**
 * Query executor
 * 
 * Handles prepared statement execution with both positional and named parameters.
 */

import { PoolClient } from 'pg';
import { StratusPool } from './pool';
import { ParamValue, QueryResult, SingleResult, ExecuteOptions } from './types';
import { detectParamStyle, parseNamedParams, validateParamCount, toPgValue } from './params';

/**
 * Query metadata for prepared statements
 */
export interface QueryInfo {
  /** Original SQL */
  sql: string;
  /** Parameter count */
  paramCount: number;
  /** Return type */
  returnType: 'one' | 'many' | 'exec';
  /** Parameter names (for debugging) */
  paramNames?: string[];
}

/**
 * Create a query executor bound to a pool
 * 
 * @param pool - StratusPool instance
 * @returns StratusExecutor
 * 
 * @example
 * ```typescript
 * import { createPool, execute } from '@stratusdb/sdk';
 * 
 * const pool = createPool(process.env.DATABASE_URL!);
 * const exec = execute(pool);
 * 
 * // Positional parameters (default - fastest)
 * const users = await exec.query({
 *   sql: 'SELECT * FROM users WHERE id = $1',
 *   params: [1],
 *   returnType: 'one',
 * });
 * 
 * // Named parameters (more readable)
 * const users = await exec.query({
 *   sql: 'SELECT * FROM users WHERE id = :id',
 *   params: { id: 1 },
 *   returnType: 'one',
 * });
 * ```
 */
export function execute(pool: StratusPool): StratusExecutor {
  return new StratusExecutor(pool);
}

/**
 * Execute a single query (convenience function)
 * 
 * @param pool - Pool instance
 * @param sql - SQL query
 * @param params - Parameters (positional array or named object)
 * @returns Query result
 */
export async function executeQuery<T = unknown>(
  pool: StratusPool,
  sql: string,
  params?: ParamValue[] | Record<string, ParamValue>,
  options?: ExecuteOptions
): Promise<QueryResult<T>> {
  const executor = new StratusExecutor(pool);
  return executor.query<T>({
    sql,
    params: params ?? [],
    returnType: 'many',
    options,
  });
}

/**
 * Execute and return single row
 * 
 * @param pool - Pool instance
 * @param sql - SQL query
 * @param params - Parameters
 * @returns Single row result
 */
export async function executeOne<T = unknown>(
  pool: StratusPool,
  sql: string,
  params?: ParamValue[] | Record<string, ParamValue>,
  options?: ExecuteOptions
): Promise<SingleResult<T>> {
  const executor = new StratusExecutor(pool);
  return executor.queryOne<T>({
    sql,
    params: params ?? [],
    returnType: 'one',
    options,
  });
}

/**
 * Stratus Query Executor
 */
export class StratusExecutor {
  private pool: StratusPool;
  
  constructor(pool: StratusPool) {
    this.pool = pool;
  }
  
  /**
   * Execute a query
   * 
   * @param query - Query info object
   * @returns Query result
   */
  async query<T = unknown>(query: {
    sql: string;
    params: ParamValue[] | Record<string, ParamValue>;
    returnType: 'one' | 'many' | 'exec';
    options?: ExecuteOptions;
  }): Promise<QueryResult<T>> {
    const { sql, params, returnType, options } = query;
    const paramStyle = detectParamStyle(sql);
    
    // Prepare parameters based on style
    let paramArray: ParamValue[];
    if (paramStyle.style === 'named') {
      paramArray = parseNamedParams(sql, params as Record<string, ParamValue>);
    } else {
      paramArray = params as ParamValue[];
    }
    
    // Validate parameter count
    if (paramStyle.paramCount > 0) {
      validateParamCount(sql, paramArray, paramStyle.paramCount);
    }
    
    // Execute query
    const result = await this.pool.query<T>(
      sql,
      paramArray.map(toPgValue)
    );
    
    // Validate result for :one queries
    if (returnType === 'one' && result.rowCount > 1) {
      console.warn(
        `Query returned ${result.rowCount} rows but expected 1. ` +
        `Use :many for multiple rows.`
      );
    }
    
    return result;
  }
  
  /**
   * Execute and return single row
   * 
   * @param query - Query info object
   * @returns Single row result
   */
  async queryOne<T = unknown>(query: {
    sql: string;
    params: ParamValue[] | Record<string, ParamValue>;
    options?: ExecuteOptions;
  }): Promise<SingleResult<T>> {
    const result = await this.query<T>({
      ...query,
      returnType: 'one',
    });
    
    return {
      row: result.rows[0] ?? null,
      found: result.rowCount > 0,
    };
  }
  
  /**
   * Execute without returning rows (INSERT, UPDATE, DELETE)
   * 
   * @param query - Query info object
   * @returns Execution result
   */
  async exec(query: {
    sql: string;
    params: ParamValue[] | Record<string, ParamValue>;
    options?: ExecuteOptions;
  }): Promise<QueryResult<null>> {
    return this.query<null>({
      ...query,
      returnType: 'exec',
    });
  }
  
  /**
   * Execute with a dedicated client (for transactions)
   * 
   * @param query - Query info object
   * @param client - Pool client
   * @returns Query result
   */
  async queryWithClient<T = unknown>(
    query: {
      sql: string;
      params: ParamValue[] | Record<string, ParamValue>;
      returnType: 'one' | 'many' | 'exec';
    },
    client: PoolClient
  ): Promise<QueryResult<T>> {
    const { sql, params, returnType } = query;
    const paramStyle = detectParamStyle(sql);
    
    let paramArray: ParamValue[];
    if (paramStyle.style === 'named') {
      paramArray = parseNamedParams(sql, params as Record<string, ParamValue>);
    } else {
      paramArray = params as ParamValue[];
    }
    
    const result = await client.query<T>(sql, paramArray.map(toPgValue));
    
    return {
      rows: result.rows,
      rowCount: result.rowCount ?? 0,
      command: result.command,
    };
  }
}

File: sdk/py/stratus/executor.py
"""
Query executor

Handles prepared statement execution with both positional and named parameters.
"""

from typing import TypeVar, Generic, Dict, Any, List, Optional, Union
from dataclasses import dataclass

from .pool import StratusPool, PoolConnection
from .types import QueryResult, SingleResult, ExecuteOptions
from .params import (
    detect_param_style,
    parse_named_params,
    validate_param_count,
    prepare_params,
)


T = TypeVar("T")


@dataclass
class QueryInfo:
    """Query metadata"""

    sql: str
    param_count: int
    return_type: str
    param_names: Optional[List[str]] = None


def execute(pool: StratusPool) -> "StratusExecutor":
    """
    Create a query executor bound to a pool

    Args:
        pool: StratusPool instance

    Returns:
        StratusExecutor

    Example:
        ```python
        from stratus import create_pool, execute

        pool = create_pool("postgresql://...")
        exec = execute(pool)

        # Positional parameters (default - fastest)
        users = await exec.query(
            sql="SELECT * FROM users WHERE id = $1",
            params=[1],
            return_type="one",
        )

        # Named parameters (more readable)
        users = await exec.query(
            sql="SELECT * FROM users WHERE id = :id",
            params={"id": 1},
            return_type="one",
        )
        ```
    """
    return StratusExecutor(pool)


async def execute_query(
    pool: StratusPool,
    sql: str,
    params: Optional[Union[List[Any], Dict[str, Any]]] = None,
    *,
    return_type: str = "many",
    options: Optional[ExecuteOptions] = None,
) -> QueryResult:
    """
    Execute a single query (convenience function)

    Args:
        pool: Pool instance
        sql: SQL query
        params: Parameters (positional list or named dict)
        return_type: 'one', 'many', or 'exec'
        options: Execution options

    Returns:
        Query result
    """
    executor = StratusExecutor(pool)
    return await executor.query(
        sql=sql,
        params=params or [],
        return_type=return_type,
        options=options,
    )


async def execute_one(
    pool: StratusPool,
    sql: str,
    params: Optional[Union[List[Any], Dict[str, Any]]] = None,
    *,
    options: Optional[ExecuteOptions] = None,
) -> SingleResult:
    """
    Execute and return single row

    Args:
        pool: Pool instance
        sql: SQL query
        params: Parameters
        options: Execution options

    Returns:
        Single row result
    """
    executor = StratusExecutor(pool)
    return await executor.query_one(
        sql=sql,
        params=params or [],
        options=options,
    )


class StratusExecutor:
    """
    Stratus Query Executor

    Handles query execution with proper parameter handling.
    """

    def __init__(self, pool: StratusPool):
        self.pool = pool

    async def query(
        self,
        sql: str,
        params: Union[List[Any], Dict[str, Any]],
        *,
        return_type: str = "many",
        options: Optional[ExecuteOptions] = None,
    ) -> QueryResult:
        """
        Execute a query

        Args:
            sql: SQL query
            params: Parameters (positional list or named dict)
            return_type: 'one', 'many', or 'exec'
            options: Execution options

        Returns:
            Query result
        """
        param_style = detect_param_style(sql)
        param_array = prepare_params(sql, params)

        if param_style[0] != "unknown" and param_style[2] > 0:
            validate_param_count(sql, param_array, param_style[2])

        async with self.pool.connect() as conn:
            result = await conn.execute(sql, param_array)

            if return_type == "one" and result.row_count > 1:
                import warnings

                warnings.warn(
                    f"Query returned {result.row_count} rows but expected 1. "
                    f"Use return_type='many' for multiple rows."
                )

            return result

    async def query_one(
        self,
        sql: str,
        params: Union[List[Any], Dict[str, Any]],
        *,
        options: Optional[ExecuteOptions] = None,
    ) -> SingleResult:
        """
        Execute and return single row

        Args:
            sql: SQL query
            params: Parameters
            options: Execution options

        Returns:
            Single row result
        """
        result = await self.query(
            sql=sql,
            params=params,
            return_type="one",
            options=options,
        )

        return SingleResult(
            row=result.rows[0] if result.rows else None,
            found=result.row_count > 0,
        )

    async def exec(
        self,
        sql: str,
        params: Union[List[Any], Dict[str, Any]],
        *,
        options: Optional[ExecuteOptions] = None,
    ) -> QueryResult:
        """
        Execute without returning rows (INSERT, UPDATE, DELETE)

        Args:
            sql: SQL query
            params: Parameters
            options: Execution options

        Returns:
            Execution result
        """
        return await self.query(
            sql=sql,
            params=params,
            return_type="exec",
            options=options,
        )

    async def query_with_connection(
        self,
        sql: str,
        params: Union[List[Any], Dict[str, Any]],
        conn: PoolConnection,
        *,
        return_type: str = "many",
    ) -> QueryResult:
        """
        Execute with a dedicated connection (for transactions)

        Args:
            sql: SQL query
            params: Parameters
            conn: Pool connection
            return_type: 'one', 'many', or 'exec'

        Returns:
            Query result
        """
        param_style = detect_param_style(sql)
        param_array = prepare_params(sql, params)

        return await conn.execute(sql, param_array)
